internal/ignition/builder.go

```diff
@@ -0,0 +1,452 @@
+          "contents": { "source": "data:text/plain;base64,{{.SELINUX_POLICY}}" }
+      }{{if .RH_ROOT_CA}},
+      {
+        "overwrite": true,
+        "path": "/etc/pki/ca-trust/source/anchors/rh-it-root-ca.crt",
+        "mode": 420,
+        "user": {
+            "name": "root"
+        },
+        "contents": { "source": "data:,{{.RH_ROOT_CA}}" }
+      }{{end}}{{if .HostCACertPath}},
+      {
+        "path": "{{.HostCACertPath}}",
+        "mode": 420,
+        "overwrite": true,
+        "user": {
+          "name": "root"
+        },
+        "contents": { "source": "{{.ServiceCACertData}}" }
+      }{{end}}{{if .ServiceIPs}},
+      {
+        "path": "/etc/hosts",
+        "mode": 420,
+        "user": {
+          "name": "root"
+        },
+        "append": [{ "source": "{{.ServiceIPs}}" }]
+      }{{end}}{{if .StaticNetworkConfig}},
+      {
+          "path": "/usr/local/bin/pre-network-manager-config.sh",
+          "mode": 493,
+          "overwrite": true,
+          "user": {
+              "name": "root"
+          },
+          "contents": { "source": "data:text/plain;base64,{{.PreNetworkConfigScript}}"}
+      }{{end}}{{range .StaticNetworkConfig}},
+      {
+        "path": "{{.FilePath}}",
+        "mode": 384,
+        "overwrite": true,
+        "user": {
+          "name": "root"
+        },
+        "contents": { "source": "data:text/plain;base64,{{.FileContents}}"}
+      }{{end}}
+    ]
+  }
+}`
+
+const nodeIgnitionFormat = `
+{
+  "ignition": {
+    "version": "3.1.0",
+    "config": {
+      "merge": [
+        {
+          "source": "{{.SOURCE}}"
+        }
+      ]
+    }
+  }
+}`
+
+const tempNMConnectionsDir = "/etc/assisted/network"
+
+// Config contains the parameters required to build the ignition file
+type IgnitionConfig struct {
+	AgentDockerImg       string        `envconfig:"AGENT_DOCKER_IMAGE" default:"quay.io/ocpmetal/assisted-installer-agent:latest"`
+	AgentTimeoutStart    time.Duration `envconfig:"AGENT_TIMEOUT_START" default:"3m"`
+	InstallRHCa          bool          `envconfig:"INSTALL_RH_CA" default:"false"`
+	ServiceBaseURL       string        `envconfig:"SERVICE_BASE_URL"`
+	ServiceCACertPath    string        `envconfig:"SERVICE_CA_CERT_PATH" default:""`
+	ServiceIPs           string        `envconfig:"SERVICE_IPS" default:""`
+	SkipCertVerification bool          `envconfig:"SKIP_CERT_VERIFICATION" default:"false"`
+}
+
+type ignitionBuilder struct {
+	staticNetworkConfig staticnetworkconfig.StaticNetworkConfig
+	cluster             *common.Cluster
+	config              IgnitionConfig
+	imageCreateParams   *models.ImageCreateParams
+	logger              logrus.FieldLogger
+	authType            auth.AuthType
+	safeForLogs         bool
+}
+
+// NewBuilder returns ignition builder which builds the ignition file based on the paramters provided
+// to it as functional options during build
+func NewBuilder(staticNetworkConfig staticnetworkconfig.StaticNetworkConfig) *ignitionBuilder {
```

Comment:
isn't it a little bit of overkill? Why not just define an API that receives 7 parameters?

Reply:
Having 7 parameters seems already too much.
By using this pattern we benefit of the following:
- No need to change the function signature (unlike with a single function) when new ignition parameter will be needed, just to provide another option to the Build method by the client.
- No need to remember the parameters order 
- Each option function can perform its own validation on the provided value (requires changing the signature to return error). The alternative is a long block of validation for each method parameter
- SafeForLog is 'false' by default, therefore only a single client of the builder will have to provide it as an option to set its value to 'true' (meaning the API can be called with a varied number of arguments)